File: tradepilot/optimization.py
# tradepilot/optimization.py
import numpy as np
from scipy.optimize import minimize as scipy_minimize
from .metrics import portfolio_return, portfolio_vol


def msr(riskfree_rate, er, cov, min_w=0.01, max_w=0.95):
    """
    Computes the optimal asset weights that maximize the Sharpe ratio.

    Parameters:
        riskfree_rate (float): The risk-free rate.
        er (pd.Series or np.array): Expected returns.
        cov (pd.DataFrame or np.array): Covariance matrix.
        min_w (float): Minimum weight for any single asset.
        max_w (float): Maximum weight for any single asset.

    Returns:
        np.array: Optimal weights vector (normalized to sum to 1).
    """
    n = len(er)
    init_guess = np.repeat(1/n, n)
    bounds = ((min_w, max_w),) * n
    weights_sum_to_1 = {'type': 'eq', 'fun': lambda weights: np.sum(weights) - 1}

    def neg_sharpe(weights, riskfree_rate, er, cov):
        r = portfolio_return(weights, er)
        vol = portfolio_vol(weights, cov)
        return -(r - riskfree_rate) / vol

    result = scipy_minimize(neg_sharpe, init_guess,
                            args=(riskfree_rate, er, cov), method='SLSQP',
                            options={'disp': False},
                            constraints=(weights_sum_to_1,),
                            bounds=bounds)
    return result.x / result.x.sum()


def gmv(cov, min_w=0.01, max_w=0.95):
    """
    Computes the weights for the Global Minimum Variance portfolio.
    Uses the MSR trick: when all expected returns are equal,
    maximizing Sharpe reduces to minimizing volatility.

    Parameters:
        cov (pd.DataFrame or np.array): Covariance matrix.
        min_w (float): Minimum weight for any single asset.
        max_w (float): Maximum weight for any single asset.

    Returns:
        np.array: Weights vector for the minimum variance portfolio.
    """
    n = cov.shape[0]
    return msr(0, np.repeat(1, n), cov, min_w=min_w, max_w=max_w)


def eq_weighted(er):
    """
    Returns an equally weighted portfolio.

    Parameters:
        er (pd.Series or np.array): Expected returns (used only to determine length).

    Returns:
        np.array: Equal weights vector.
    """
    n = len(er)
    return np.repeat(1/n, n)


def minimize_vol(target_return, er, cov):
    """
    Returns the optimal weights that achieve the target return
    with minimum volatility, given expected returns and a covariance matrix.

    Parameters:
        target_return (float): The target portfolio return.
        er (pd.Series or np.array): Expected returns.
        cov (pd.DataFrame or np.array): Covariance matrix.

    Returns:
        np.array: Optimal weights that minimize volatility for the target return.
    """
    n = er.shape[0]
    init_guess = np.repeat(1/n, n)
    bounds = ((0.00, 1),) * n
    weights_sum_to_1 = {
        'type': 'eq',
        'fun': lambda weights: np.sum(weights) - 1
    }
    return_is_target = {
        'type': 'eq',
        'args': (er,),
        'fun': lambda weights, er: target_return - portfolio_return(weights, er)
    }
    result = scipy_minimize(portfolio_vol, init_guess,
                            args=(cov,), method='SLSQP',
                            options={'disp': False},
                            constraints=(weights_sum_to_1, return_is_target),
                            bounds=bounds)
    return result.x


def optimal_weights(n_points, er, cov):
    """
    Returns a list of weights that represent a grid of n_points
    on the efficient frontier.

    Parameters:
        n_points (int): Number of points on the efficient frontier.
        er (pd.Series or np.array): Expected returns.
        cov (pd.DataFrame or np.array): Covariance matrix.

    Returns:
        list[np.array]: List of weight arrays along the efficient frontier.
    """
    target_rs = np.linspace(er.min(), er.max(), n_points)
    weights = [minimize_vol(target_return, er, cov) for target_return in target_rs]
    return weights

File: web/src/lib/engine/optimization.ts
/**
 * TradePilot Engine — Portfolio Optimization
 *
 * Pure TypeScript port of tradepilot/optimization.py.
 * Implements MSR, GMV, EW, and efficient frontier computation.
 *
 * For constrained optimization (scipy.optimize.minimize equivalent),
 * we implement a projected gradient descent with box + equality constraints.
 */

import { portfolioReturn, portfolioVol } from './metrics';
import type { EfficientFrontierResult } from './types';

// ---------------------------------------------------------------------------
// Constrained optimizer (replaces scipy.optimize.minimize SLSQP)
// ---------------------------------------------------------------------------

/**
 * Project weights onto the feasible set:
 * - Clip to [minW, maxW]
 * - Normalize so sum = 1
 *
 * Uses iterative clipping + rescaling (converges quickly for small N).
 */
function projectWeights(w: number[], minW: number, maxW: number): number[] {
  const n = w.length;
  const out = new Float64Array(n);
  for (let i = 0; i < n; i++) out[i] = w[i];

  for (let iter = 0; iter < 50; iter++) {
    // Clip
    for (let i = 0; i < n; i++) {
      out[i] = Math.max(minW, Math.min(maxW, out[i]));
    }
    // Normalize
    let sum = 0;
    for (let i = 0; i < n; i++) sum += out[i];
    if (sum === 0) {
      for (let i = 0; i < n; i++) out[i] = 1 / n;
      break;
    }
    for (let i = 0; i < n; i++) out[i] /= sum;

    // Check feasibility
    let feasible = true;
    for (let i = 0; i < n; i++) {
      if (out[i] < minW - 1e-10 || out[i] > maxW + 1e-10) {
        feasible = false;
        break;
      }
    }
    if (feasible) break;
  }

  return Array.from(out);
}

/**
 * Numerical gradient of f with respect to weights.
 */
function numericalGradient(
  f: (w: number[]) => number,
  w: number[],
  eps = 1e-7,
): number[] {
  const n = w.length;
  const grad: number[] = new Array(n);
  const f0 = f(w);
  for (let i = 0; i < n; i++) {
    const wPlus = [...w];
    wPlus[i] += eps;
    grad[i] = (f(wPlus) - f0) / eps;
  }
  return grad;
}

/**
 * Minimise an objective function subject to:
 * - weights sum to 1
 * - minW <= w_i <= maxW
 *
 * Uses projected gradient descent with adaptive step size.
 */
function minimizeConstrained(
  objective: (w: number[]) => number,
  n: number,
  minW: number,
  maxW: number,
  maxIter = 500,
  tol = 1e-10,
): number[] {
  let w = new Array(n).fill(1 / n);
  let lr = 0.1;
  let prevObj = objective(w);

  for (let iter = 0; iter < maxIter; iter++) {
    const grad = numericalGradient(objective, w);

    // Step
    const wNew = w.map((wi, i) => wi - lr * grad[i]);

    // Project
    const wProj = projectWeights(wNew, minW, maxW);
    const newObj = objective(wProj);

    if (newObj < prevObj) {
      w = wProj;
      const improvement = Math.abs(prevObj - newObj);
      prevObj = newObj;
      lr *= 1.05; // accelerate
      if (improvement < tol) break;
    } else {
      lr *= 0.5; // shrink step
      if (lr < 1e-15) break;
    }
  }

  return projectWeights(w, minW, maxW);
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Maximum Sharpe Ratio (MSR) portfolio.
 *
 * max_w (w^T * mu - rf) / sqrt(w^T * Sigma * w)
 * s.t.  sum(w) = 1, minW <= w_i <= maxW
 *
 * @param riskFreeRate Annual risk-free rate.
 * @param expectedReturns Expected return per asset (annualized).
 * @param covMatrix Covariance matrix of returns.
 * @param minW Minimum weight per asset (default 0.01).
 * @param maxW Maximum weight per asset (default 0.95).
 * @returns Optimal weight vector.
 */
export function msr(
  riskFreeRate: number,
  expectedReturns: number[],
  covMatrix: number[][],
  minW = 0.01,
  maxW = 0.95,
): number[] {
  const n = expectedReturns.length;
  if (n === 0) return [];
  if (n === 1) return [1];

  const negSharpe = (w: number[]): number => {
    const ret = portfolioReturn(w, expectedReturns);
    const vol = portfolioVol(w, covMatrix);
    if (vol < 1e-15) return 1e10;
    return -(ret - riskFreeRate) / vol;
  };

  return minimizeConstrained(negSharpe, n, minW, maxW);
}

/**
 * Global Minimum Variance (GMV) portfolio.
 *
 * min_w sqrt(w^T * Sigma * w)
 *
 * Uses the MSR trick: when all expected returns are equal,
 * maximising Sharpe = minimising variance.
 *
 * @param covMatrix Covariance matrix.
 * @param minW Minimum weight per asset.
 * @param maxW Maximum weight per asset.
 * @returns Minimum variance weight vector.
 */
export function gmv(
  covMatrix: number[][],
  minW = 0.01,
  maxW = 0.95,
): number[] {
  const n = covMatrix.length;
  const equalReturns = new Array(n).fill(1);
  return msr(0, equalReturns, covMatrix, minW, maxW);
}

/**
 * Equal Weight portfolio.
 *
 * w_i = 1/N for all i.
 *
 * @param n Number of assets.
 * @returns Equal weight vector.
 */
export function equalWeight(n: number): number[] {
  if (n <= 0) return [];
  return new Array(n).fill(1 / n);
}

/**
 * Minimise portfolio volatility for a target return.
 * Used to trace the efficient frontier.
 */
export function minimizeVol(
  targetReturn: number,
  expectedReturns: number[],
  covMatrix: number[][],
): number[] {
  const n = expectedReturns.length;

  const objective = (w: number[]): number => {
    const vol = portfolioVol(w, covMatrix);
    // Penalty for deviating from target return
    const retDiff = portfolioReturn(w, expectedReturns) - targetReturn;
    return vol + 100 * retDiff * retDiff;
  };

  return minimizeConstrained(objective, n, 0, 1, 1000, 1e-12);
}

/**
 * Compute the efficient frontier: a grid of portfolios
 * from the minimum-return asset to the maximum-return asset.
 *
 * @param expectedReturns Annualized expected returns per asset.
 * @param covMatrix Covariance matrix.
 * @param nPoints Number of points on the frontier.
 * @returns Arrays of returns, volatilities, and weight vectors.
 */
export function efficientFrontier(
  expectedReturns: number[],
  covMatrix: number[][],
  nPoints = 50,
): EfficientFrontierResult {
  const minR = Math.min(...expectedReturns);
  const maxR = Math.max(...expectedReturns);

  const rets: number[] = [];
  const vols: number[] = [];
  const weights: number[][] = [];

  for (let i = 0; i < nPoints; i++) {
    const targetRet = minR + (maxR - minR) * (i / (nPoints - 1));
    const w = minimizeVol(targetRet, expectedReturns, covMatrix);
    const actualRet = portfolioReturn(w, expectedReturns);
    const vol = portfolioVol(w, covMatrix);
    rets.push(actualRet);
    vols.push(vol);
    weights.push(w);
  }

  return { returns: rets, vols, weights };
}
